doc/BGP/BGP-supress-fib-pending.md

```diff
@@ -0,0 +1,804 @@
+> - The feature will not apply for routes learnt through other means like redistribution to bgp from other protocols. This is applicable only to peer learnt routes.
+> - If a route is installed in FIB and then gets deleted from the dataplane, then routes will not be withdrawn from peers. This will be considered as dataplane issue.
+> - The feature will slightly increase the time required to advertise the routes to peers since the route install status needs to be received from the FIB
+> - If routes are received by the peer before the configuration is applied, then the bgp sessions need to be reset for the configuration to take effect.
+> - If the route which is already installed in dataplane is removed for some reason, sending withdraw message to peers is not currently supported.
+
+BGP will queue routes which aren't installed in the FIB. To make zebra install routes in FIB an RTM_NEWROUTE message with *RTM_F_OFFLOAD* flag set must be sent by SONiC to zebra through FPM socket.
+
+**NOTE**: This feature is implemented as part of new *dplane_fpm_nl* zebra plugin in FRR 8.4 and a backport patch must be created for current FRR 8.2 SONiC is using. SONiC is still using old *fpm* plugin which isn't developed anymore and thus SONiC must migrate to the new implementation as part of this change.
+
+### 4. Requirements
+
+High level requirements:
+
+- Provide an end to end ASIC hardware to BGP route programming feedback mechanism
+- BGP must not advertise routes which aren't yet offloaded if this feature is enabled
+- BGP route suppression is optional and disabled by default
+- Route suppression can be turned on or off at runtime
+- Offloaded routes consistency monitoring and automatic mitigation
+
+Restrictions/Limitations:
+
+- MPLS, VNET routes are out of scope of this document
+- Directly connected and static routes are announced by BGP regardless of their offload status
+
+### 5. Architecture Design
+
+<!-- omit in toc -->
+##### Figure 2. Architecture
+
+The below diagram shows the high level architecture including sending an RTM_NEWROUTE response message with RTM_F_OFFLOAD flag set to zebra through FPM:
+
+<p align=center>
+<img src="img/architecture-diagram.png" alt="Figure 2. Architecture diagram">
+</p>
+
+- *2*-*8* - existing flows in SONiC
+- *1*, *9*, *10* - new flows to be added
+
+### 6. Configuration and management
+
+#### 6.1. Config DB Enhancements
+
+<!-- omit in toc -->
+##### 6.1.1.DEVICE_METADATA
+
+Configuration schema in ABNF format:
+
+```abnf
+; DEVICE_METADATA table
+key                      = DEVICE_METADATA|localhost ; Device metadata configuration table
+suppress-fib-pending     = "enabled"/"disabled"      ; Globally enable/disable routes announcement suppression feature,
```

Comment:
If this controls only for BGP routes, please be explicit in the field name.

Reply:
It is implemented in fpmsyncd/zebra which does not care if it receives BGP or other protocol routes, so, if SONiC supports, the same feature could be applied to other routing protocols.

Reply:
Since we talk only about BGP routes in the HLD, I thought it is better to have a name like "suppress-bgp-routes-fib-pending" and handle only BGP routes case. If you think, that's not the case and you'll enable support for all route types (BGP, Static, OSPF..etc), we can have the generic field name.